index.js

```diff
@@ -63,34 +113,123 @@ module.exports = () => {
 
     selectIpfsReady: state => state.ipfs.ready,
 
-    selectIpfsApiAddress: state => state.ipfs.apiAddress,
+    selectIpfsProvider: state => state.ipfs.provider,
+
+    selectIpfsApiOpts: state => state.ipfs.apiOpts,
 
-    selectIpfsInitFailed: state => !!state.ipfs.error,
+    selectIpfsInitFailed: state => state.ipfs.failed,
 
     selectIpfsIdentity: state => state.ipfs.identity,
 
-    doInitIpfs: () => async ({ dispatch, getState }) => {
-      dispatch({ type: 'IPFS_INIT_STARTED' })
+    doInitIpfs: () => async (store) => {
+      await getIpfs(opts, store)
+    },
 
-      const { apiAddress } = getState().ipfs
-      let identity = null
+    doUpdateIpfsApiOpts: (usrOpts) => (store) => {
+      store.dispatch({ type: 'IPFS_API_UPDATED', payload: usrOpts })
+      saveUserOpts('ipfsApi', usrOpts)
 
-      try {
-        ipfs = await getIpfs({ api: true, ipfs: apiAddress })
-        // will fail if remote api is not available on default port
-        identity = await ipfs.id()
-        // if it works, save the address for the next time
-        persistAddress(apiAddress)
-      } catch (error) {
-        return dispatch({ type: 'IPFS_INIT_FAILED', error })
-      }
+      getIpfs(Object.assign({}, opts, {
+        tryWindow: false,
+        tryJsIpfs: false
+      }), store)
+    }
+  }
+}
 
-      dispatch({ type: 'IPFS_INIT_FINISHED', payload: identity })
-    },
+async function tryWindow () {
+  console.log('Trying window.ipfs')
+
+  // Opportunistic optimizations when running from ipfs-companion (+ ipfs-desktop in future)
+  if (typeof browser === 'object') {
+    // Note: under some vendors getBackgroundPage() will return null if window is in incognito mode
+    const webExt = await browser.runtime.getBackgroundPage()
+    // If extension is ipfs-companion exposing IPFS API, use it directly for best performance
+    if (webExt && webExt.ipfsCompanion && webExt.ipfsCompanion.ipfs) {
+      const ipfs = webExt.ipfsCompanion.ipfs
+      const identity = await ipfs.id()
+
+      return { identity, ipfs }
+    }
+  }
 
-    doUpdateIpfsAPIAddress: (apiAddress) => ({dispatch, store}) => {
-      dispatch({type: 'IPFS_API_UPDATED', payload: apiAddress})
-      store.doInitIpfs()
+  if (root.ipfs) {
+    try {
+      let identity = await root.ipfs.id()
+      console.log('Found `window.ipfs`. Nice!')
+
+      return { ipfs: root.ipfs, identity }
+    } catch (error) {
+      console.log('Failed to get id from window.ipfs', error)
     }
+  } else {
+    console.log('No window.ipfs found. Consider Installing the IPFS Companion web extension - https://github.com/ipfs-shipyard/ipfs-companion')
+  }
+}
+
+async function tryApi (opts) {
+  try {
+    console.time('IPFS_INIT_API')
```

Comment:
I got carried away with the logging here, we can merge the timer logs and the "ready!" console message into one , by removing `console.time('IPFS_INIT_API')` and doing:

```js
console.time('js-ipfs-api ready!')
// ... stuff and magic
console.timeEnd('js-ipfs-api ready!')
```